### PR TITLE
[Enhancement] Support asynchronous profile processes

### PR DESCRIPTION
**Pull request by liuyehcf**
## Background

The total query time comprises of the following parts:
1. plan
1. execution
1. profile collection
1. profile process

In some scenerios, some operators fail to quickly response to short circuit, the time difference between fragment instances may vary significantly, which thereby lead to long profile collection.

And in some other scenerios where the query is super complex, the profile will also become super large, and the profile process time may raise to nearly 500ms.

But the client needn't to wait the process to be finished(part 3 and part 4), because the result has been already generated and transmitted. So we can async profile process to reduce the client waiting time.

## Main work of this pr

Async the profile collect and profile process steps, and decouple these to steps from the query execution process.
1. Add listener mechanism to `MarkedCountDownLatch`
2. Setup a separate thread pool, with 2 threads by default, to handle the profile process.
3. Introduce a monitor mechanism of `QeProcessorImpl` to gracefully maintain the context of query.

And the process will be executed like this:

```
generate_plan();
execution();
wait_result();

if(size(profile_thread_queue) > Config.profile_process_blocking_queue_size) {
      collect_profile_sync();
      process_profile();
} else {
      Task profile_process_task = { process_profile(); }
      add_listener_to_profile_collect(profile_process_task);
}
```

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:

- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:disappointed: **fail** : 17 / 52 (32.69%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/load/loadv2/LoadLoadingTask.java|0|1|00.00%|[219]|
|:large_blue_circle:|com/starrocks/task/ExportExportingTask.java|0|1|00.00%|[376]|
|:large_blue_circle:|com/starrocks/qe/QeProcessorImpl.java|4|24|16.67%|[111, 112, 113, 114, 115, 116, 118, 119, 120, 121, 122, 123, 124, 125, 126, 127, 128, 129, 205, 206]|
|:large_blue_circle:|com/starrocks/qe/StmtExecutor.java|8|21|38.10%|[623, 624, 626, 780, 781, 782, 785, 787, 788, 789, 1668, 1684, 1687]|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/common/Config.java|2|2|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:disappointed: **fail** : 37 / 103 (35.92%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/load/loadv2/LoadLoadingTask.java|0|1|00.00%|[219]|
|:large_blue_circle:|com/starrocks/sql/ExplainAnalyzer.java|0|5|00.00%|[365, 366, 367, 368, 370]|
|:large_blue_circle:|com/starrocks/task/ExportExportingTask.java|0|1|00.00%|[376]|
|:large_blue_circle:|com/starrocks/load/streamload/StreamLoadTask.java|0|1|00.00%|[1060]|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|1|4|25.00%|[2114, 2118, 2119]|
|:large_blue_circle:|com/starrocks/qe/StmtExecutor.java|8|32|25.00%|[333, 615, 623, 624, 781, 782, 784, 785, 786, 787, 788, 789, 791, 792, 793, 794, 795, 796, 797, 920, 1677, 1684, 1685, 1703]|
|:large_blue_circle:|com/starrocks/qe/DefaultCoordinator.java|5|18|27.78%|[421, 949, 950, 952, 953, 955, 956, 957, 959, 960, 961, 962, 964]|
|:large_blue_circle:|com/starrocks/qe/scheduler/QueryRuntimeProfile.java|2|6|33.33%|[221, 222, 225, 226]|
|:large_blue_circle:|com/starrocks/common/util/concurrent/MarkedCountDownLatch.java|8|15|53.33%|[65, 66, 110, 111, 112, 113, 114]|
|:large_blue_circle:|com/starrocks/qe/QeProcessorImpl.java|10|17|58.82%|[114, 115, 116, 118, 124, 125, 129]|
|:large_blue_circle:|com/starrocks/scheduler/PartitionBasedMvRefreshProcessor.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/common/Config.java|2|2|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:disappointed: **fail** : 49 / 108 (45.37%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/load/loadv2/LoadLoadingTask.java|0|1|00.00%|[219]|
|:large_blue_circle:|com/starrocks/sql/ExplainAnalyzer.java|0|5|00.00%|[365, 366, 367, 368, 370]|
|:large_blue_circle:|com/starrocks/task/ExportExportingTask.java|0|1|00.00%|[376]|
|:large_blue_circle:|com/starrocks/load/streamload/StreamLoadTask.java|0|1|00.00%|[1060]|
|:large_blue_circle:|com/starrocks/qe/StmtExecutor.java|8|35|22.86%|[333, 615, 623, 624, 781, 782, 786, 788, 789, 790, 793, 794, 795, 796, 798, 799, 800, 801, 803, 804, 805, 806, 929, 1686, 1693, 1694, 1712]|
|:large_blue_circle:|com/starrocks/qe/scheduler/QueryRuntimeProfile.java|2|6|33.33%|[221, 222, 225, 226]|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|2|4|50.00%|[2118, 2119]|
|:large_blue_circle:|com/starrocks/common/util/concurrent/MarkedCountDownLatch.java|8|15|53.33%|[65, 66, 110, 111, 112, 113, 114]|
|:large_blue_circle:|com/starrocks/qe/QeProcessorImpl.java|12|19|63.16%|[114, 115, 116, 117, 119, 125, 126]|
|:large_blue_circle:|com/starrocks/qe/DefaultCoordinator.java|14|18|77.78%|[421, 950, 953, 964]|
|:large_blue_circle:|com/starrocks/scheduler/PartitionBasedMvRefreshProcessor.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/common/Config.java|2|2|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:disappointed: **fail** : 50 / 117 (42.74%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/load/loadv2/LoadLoadingTask.java|0|1|00.00%|[219]|
|:large_blue_circle:|com/starrocks/sql/ExplainAnalyzer.java|0|5|00.00%|[365, 366, 367, 368, 370]|
|:large_blue_circle:|com/starrocks/task/ExportExportingTask.java|0|1|00.00%|[376]|
|:large_blue_circle:|com/starrocks/load/streamload/StreamLoadTask.java|0|1|00.00%|[1060]|
|:large_blue_circle:|com/starrocks/qe/StmtExecutor.java|8|35|22.86%|[333, 615, 623, 624, 781, 782, 786, 788, 789, 790, 793, 794, 795, 796, 798, 799, 800, 801, 803, 804, 805, 806, 929, 1686, 1693, 1694, 1712]|
|:large_blue_circle:|com/starrocks/qe/scheduler/QueryRuntimeProfile.java|2|6|33.33%|[221, 222, 225, 226]|
|:large_blue_circle:|com/starrocks/common/util/concurrent/MarkedCountDownLatch.java|9|24|37.50%|[66, 67, 68, 112, 113, 114, 115, 116, 120, 123, 124, 125, 129, 130, 132]|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|2|4|50.00%|[2118, 2119]|
|:large_blue_circle:|com/starrocks/qe/QeProcessorImpl.java|12|19|63.16%|[114, 115, 116, 117, 119, 125, 126]|
|:large_blue_circle:|com/starrocks/qe/DefaultCoordinator.java|14|18|77.78%|[421, 950, 953, 964]|
|:large_blue_circle:|com/starrocks/scheduler/PartitionBasedMvRefreshProcessor.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/common/Config.java|2|2|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:disappointed: **fail** : 40 / 125 (32.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/load/loadv2/LoadLoadingTask.java|0|1|00.00%|[219]|
|:large_blue_circle:|com/starrocks/sql/ExplainAnalyzer.java|0|5|00.00%|[365, 366, 367, 368, 370]|
|:large_blue_circle:|com/starrocks/task/ExportExportingTask.java|0|1|00.00%|[376]|
|:large_blue_circle:|com/starrocks/load/streamload/StreamLoadTask.java|0|1|00.00%|[1060]|
|:large_blue_circle:|com/starrocks/qe/StmtExecutor.java|8|41|19.51%|[616, 624, 625, 782, 783, 787, 789, 790, 791, 792, 796, 797, 798, 799, 800, 801, 804, 805, 806, 807, 809, 810, 811, 812, 814, 815, 816, 817, 940, 1697, 1704, 1705, 1723]|
|:large_blue_circle:|com/starrocks/qe/scheduler/QueryRuntimeProfile.java|2|8|25.00%|[222, 223, 226, 227, 228, 229]|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|1|4|25.00%|[2143, 2147, 2148]|
|:large_blue_circle:|com/starrocks/qe/DefaultCoordinator.java|5|18|27.78%|[422, 950, 951, 953, 954, 956, 957, 958, 960, 961, 962, 963, 965]|
|:large_blue_circle:|com/starrocks/common/util/concurrent/MarkedCountDownLatch.java|9|24|37.50%|[66, 67, 68, 112, 113, 114, 115, 116, 120, 123, 124, 125, 129, 130, 132]|
|:large_blue_circle:|com/starrocks/qe/QeProcessorImpl.java|12|19|63.16%|[114, 115, 116, 117, 119, 125, 126]|
|:large_blue_circle:|com/starrocks/scheduler/PartitionBasedMvRefreshProcessor.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/common/Config.java|2|2|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/qe/scheduler/QueryRuntimeProfile.java [hunk 1]
```diff
@@ -205,6 +217,15 @@ public boolean isFinished() {
...
+            return false;
+        }
+        // We need to make sure this submission won't be rejected by set the queue size to Integer.MAX_VALUE
+        profileDoneSignal.addListener(() -> EXECUTOR.submit(task));
+        return true;
+    }
+
     public boolean waitForProfileFinished(long timeout, TimeUnit unit) {
         boolean res = false;
         try {
```
[the remark this answers is not in the document]
Comment (liuyehcf) [reply fe/fe-core/src/main/java/com/starrocks/qe/scheduler/QueryRuntimeProfile.java:234]: Yes, it cannot guarantee that the queu size is smaller than the `Config.profile_process_blocking_queue_size`, so I set the real queue size of the thread poll to `Integer.MAX_VALUE`.

File: fe/fe-core/src/main/java/com/starrocks/common/util/concurrent/MarkedCountDownLatch.java [hunk 2]
```diff
@@ -89,5 +98,19 @@ public synchronized void countDownToZero(Status status) {
...
+        }
+        for (Runnable listener : listeners) {
+            try {
+                listener.run();
+            } catch (Throwable e) {
+                LOG.warn("Listener invoke failed", e);
+            }
+        }
     }
 }
```
[the remark this answers is not in the document]
Comment (liuyehcf) [reply fe/fe-core/src/main/java/com/starrocks/common/util/concurrent/MarkedCountDownLatch.java:134]: all the operations on the listeners is protected by `synchronized` keyword, so there no multi-thread issue here.

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:disappointed: **fail** : 71 / 129 (55.04%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/load/loadv2/LoadLoadingTask.java|0|1|00.00%|[219]|
|:large_blue_circle:|com/starrocks/sql/ExplainAnalyzer.java|0|5|00.00%|[365, 366, 367, 368, 370]|
|:large_blue_circle:|com/starrocks/task/ExportExportingTask.java|0|1|00.00%|[376]|
|:large_blue_circle:|com/starrocks/load/streamload/StreamLoadTask.java|0|1|00.00%|[1060]|
|:large_blue_circle:|com/starrocks/qe/StmtExecutor.java|8|41|19.51%|[616, 624, 625, 782, 783, 787, 789, 790, 791, 792, 796, 797, 798, 799, 800, 801, 804, 805, 806, 807, 809, 810, 811, 812, 814, 815, 816, 817, 940, 1697, 1704, 1705, 1723]|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|2|4|50.00%|[2147, 2148]|
|:large_blue_circle:|com/starrocks/qe/DefaultCoordinator.java|15|22|68.18%|[422, 951, 954, 966, 967, 968, 969]|
|:large_blue_circle:|com/starrocks/qe/QeProcessorImpl.java|14|19|73.68%|[114, 115, 116, 117, 119]|
|:large_blue_circle:|com/starrocks/qe/scheduler/QueryRuntimeProfile.java|7|8|87.50%|[223]|
|:large_blue_circle:|com/starrocks/common/util/concurrent/MarkedCountDownLatch.java|22|24|91.67%|[113, 114]|
|:large_blue_circle:|com/starrocks/scheduler/PartitionBasedMvRefreshProcessor.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/common/Config.java|2|2|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:disappointed: **fail** : 54 / 129 (41.86%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/load/loadv2/LoadLoadingTask.java|0|1|00.00%|[219]|
|:large_blue_circle:|com/starrocks/sql/ExplainAnalyzer.java|0|5|00.00%|[365, 366, 367, 368, 370]|
|:large_blue_circle:|com/starrocks/task/ExportExportingTask.java|0|1|00.00%|[376]|
|:large_blue_circle:|com/starrocks/load/streamload/StreamLoadTask.java|0|1|00.00%|[1060]|
|:large_blue_circle:|com/starrocks/qe/StmtExecutor.java|8|41|19.51%|[616, 624, 625, 782, 783, 787, 789, 790, 791, 792, 796, 797, 798, 799, 800, 801, 804, 805, 806, 807, 809, 810, 811, 812, 814, 815, 816, 817, 940, 1697, 1704, 1705, 1723]|
|:large_blue_circle:|com/starrocks/qe/scheduler/QueryRuntimeProfile.java|2|8|25.00%|[222, 223, 226, 227, 228, 229]|
|:large_blue_circle:|com/starrocks/common/util/concurrent/MarkedCountDownLatch.java|9|24|37.50%|[66, 67, 68, 112, 113, 114, 115, 116, 120, 123, 124, 125, 129, 130, 132]|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|2|4|50.00%|[2147, 2148]|
|:large_blue_circle:|com/starrocks/qe/QeProcessorImpl.java|12|19|63.16%|[114, 115, 116, 117, 119, 125, 126]|
|:large_blue_circle:|com/starrocks/qe/DefaultCoordinator.java|18|22|81.82%|[422, 951, 954, 971]|
|:large_blue_circle:|com/starrocks/scheduler/PartitionBasedMvRefreshProcessor.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/common/Config.java|2|2|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:disappointed: **fail** : 54 / 129 (41.86%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/load/loadv2/LoadLoadingTask.java|0|1|00.00%|[219]|
|:large_blue_circle:|com/starrocks/sql/ExplainAnalyzer.java|0|5|00.00%|[365, 366, 367, 368, 370]|
|:large_blue_circle:|com/starrocks/task/ExportExportingTask.java|0|1|00.00%|[376]|
|:large_blue_circle:|com/starrocks/load/streamload/StreamLoadTask.java|0|1|00.00%|[1060]|
|:large_blue_circle:|com/starrocks/qe/StmtExecutor.java|8|41|19.51%|[616, 624, 625, 782, 783, 787, 789, 790, 791, 792, 796, 797, 798, 799, 800, 801, 804, 805, 806, 807, 809, 810, 811, 812, 814, 815, 816, 817, 940, 1697, 1704, 1705, 1723]|
|:large_blue_circle:|com/starrocks/qe/scheduler/QueryRuntimeProfile.java|2|8|25.00%|[222, 223, 226, 227, 228, 229]|
|:large_blue_circle:|com/starrocks/common/util/concurrent/MarkedCountDownLatch.java|9|24|37.50%|[66, 67, 68, 112, 113, 114, 115, 116, 120, 123, 124, 125, 129, 130, 132]|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|2|4|50.00%|[2147, 2148]|
|:large_blue_circle:|com/starrocks/qe/QeProcessorImpl.java|12|19|63.16%|[114, 115, 116, 117, 119, 125, 126]|
|:large_blue_circle:|com/starrocks/qe/DefaultCoordinator.java|18|22|81.82%|[422, 951, 954, 971]|
|:large_blue_circle:|com/starrocks/scheduler/PartitionBasedMvRefreshProcessor.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/common/Config.java|2|2|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/qe/DefaultCoordinator.java [hunk 3]
```diff
@@ -942,6 +946,31 @@ public void endProfile() {
...
+        if (connectContext != null && connectContext.getSessionVariable() != null) {
+            enableAsyncProfile = connectContext.getSessionVariable().isEnableAsyncProfile();
+        }
+        TUniqueId queryId = null;
+        if (connectContext != null) {
+            queryId = connectContext.getExecutionId();
+        }
+
+        if (!enableAsyncProfile || !queryProfile.addListener(task)) {
+            LOG.warn("Profile task is full, execute in sync mode, query id = {}", DebugUtil.printId(queryId));
```
[the remark this answers is not in the document]
Comment (liuyehcf) [reply fe/fe-core/src/main/java/com/starrocks/qe/DefaultCoordinator.java:966]: Oh, that's right, I'll update.

**Review by packy92 — DISMISSED** [review]

**Review by kangkaisen — APPROVED** [review]

**Review by Seaven — APPROVED** [review]

**Comment by liuyehcf** [comment]
@Mergifyio backport branch-3.1

**Pull request merged**